### PR TITLE
Implement feature request #189 (Option to use global tslint instance)

### PR DESCRIPTION
**Pull request by smhxx**
Howdy, all! 👋 

I stumbled across #189 earlier today after wishing for the exact same feature... it's fairly simple and `linter-eslint` already has a similar option, so I decided it was worth wrangling their implementation into `linter-tslint` as well.

There were some minor linting issues with the current codebase due to recent versions of eslint (you can see what I'm talking about in the CI report for #195) so I took the initiative to resolve those as well. Two of them were super minor and well-founded, but the other one introduced by v4.6.0 was a bit silly (as others have [mentioned](https://github.com/eslint/eslint/pull/8102#issuecomment-326848403),) so I went ahead and // eslint-disabled the relevant lines to avoid mucking the style up without a second opinion. Anyway, I hope this is helpful, and any feedback is of course welcome. 😃 👍

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,33 +19,53 @@
     "rulesDirectory": {
       "type": "string",
       "title": "Custom rules directory (absolute path)",
-      "default": ""
+      "default": "",
+      "order": 2
```
Comment (Arcanemagus) [inline package.json]: If you're going to add `order` to these, re-order them in `package.json` itself to match the `order` 😉.

File: package.json [hunk 2]
```diff
@@ -19,33 +19,53 @@
...
       "title": "Fix errors on save",
       "description": "Have tslint attempt to fix some errors automatically when saving the file.",
       "type": "boolean",
-      "default": false
+      "default": false,
+      "order": 3
+    },
+    "useGlobalTslint": {
+      "type": "boolean",
+      "title": "Use the global tslint install if there is no local version (or it is disabled)",
```
Comment (Arcanemagus) [inline package.json]: This should be short and sweet: `Use the global TSLint install`. Combine the rest into the description.

**Comment by smhxx** [comment]
Alright, I'll make those tweaks tonight, then rebase and update the PR. :+1:

**Comment by smhxx** [comment]
Okay, changes made and squashed into the original commit! Should be in good shape now. 😄

File: lib/worker.js [hunk 3]
```diff
@@ -93,6 +95,22 @@ async function getLocalLinter(basedir) {
   });
 }
 
+async function getNodePrefixPath() {
```
Comment (Xapphire13) [inline lib/worker.js]: This function is marked `async`, but I don't see any async calls

File: lib/worker.js [hunk 4]
```diff
@@ -93,6 +95,22 @@ async function getLocalLinter(basedir) {
...
+  const npmCommand = process.platform === 'win32' ? 'npm.cmd' : 'npm';
+  let nodePrefixPath;
+  try {
+    nodePrefixPath = ChildProcess.spawnSync(npmCommand, ['get', 'prefix'], {
```
Comment (Xapphire13) [inline lib/worker.js]: We shouldn't use the `*Sync` versions of system commands, this takes processing time away from other atom packages due to us blocking on io

File: package.json [hunk 5]
```diff
@@ -16,36 +16,56 @@
...
-      "default": false
+      "title": "Use the global tslint install",
+      "description": "If enabled, the global tslint installation will be used as a fallback, instead of the version packaged with linter-tslint.",
+      "default": false,
+      "order": 6
+    },
+    "globalNodePath": {
+      "type": "string",
+      "title": "Global node installation path",
+      "description": "The location of your global npm install (`npm get prefix`). Optional.",
```
Comment (Xapphire13) [inline package.json]: Maybe add ``(will default to `npm get prefix`)``

File: lib/worker.js (hunk 4 above)
Comment (smhxx) [reply lib/worker.js]: Whoops, my mistake! In my haste to port over `linter-eslint`'s implementation, I forgot to adjust it for the fact that the rest of the code here is all nice and async'd. Didn't catch that they were using `spawnSync()`, although in retrospect it should have been obvious. xD I'll get right on that. 😅

**Comment by smhxx** [comment]
Alright, I went back and looked over `worker.js` and realized that I had actually made a couple mistakes there. I've gone ahead and made some corrections and squashed them into the original commit. The basics:

1. Got rid of the `spawnSync()` like you suggested, and replaced it with an `exec()`, so now the function that was mysteriously marked as `async` now actually *is* async. 😝 
2. I realized that the way I was using `getLocalLinter()` was wrong and lazy, and as a result, the global linter wasn't being cached properly, potentially leading to `npm get prefix` being run *every time* the linter was requested. (Yikes.) I had to add a second parameter to that function to get it to work correctly for both the local & global linters, but hopefully the way I did that makes sense to you guys. I also renamed it to `resolveAndCacheLinter()` to give a clearer indication of the function's purpose (especially now that it's not just being used for finding the local package.)
3. Fixed a dumb typo. :x 

I also went through with the Developer Tools up and ensured that it was using the correct linter, caching correctly, etc. and I'm now confident that it's working the way I intended. 👍

File: lib/worker.js [hunk 6]
```diff
@@ -86,13 +89,30 @@ async function getLocalLinter(basedir) {
         } else {
           linter = tslintDef;
         }
-        tslintCache.set(basedir, linter);
+        tslintCache.set(fileDir, linter);
```
Comment (Xapphire13) [inline lib/worker.js]: `linterDir`?

File: lib/worker.js [hunk 7]
```diff
@@ -86,13 +89,30 @@ async function getLocalLinter(basedir) {
...
 
+async function getNodePrefixPath() {
+  return new Promise((fulfill) => {
+    const npmCommand = process.platform === 'win32' ? 'npm.cmd' : 'npm';
+    ChildProcess.exec(
+      `${npmCommand} get prefix`,
+      { env: Object.assign(Object.assign({}, process.env), { PATH: getPath() }) },
+      (err, stdout, stderr) => {
+        if (err || stderr) {
+          throw (err || new Error(stderr));
```
Comment (Xapphire13) [inline lib/worker.js]: I haven't tried this before, does this bubble up correctly as if you used the reject of the promise?

File: lib/worker.js [hunk 8]
```diff
@@ -103,7 +123,24 @@ async function getLinter(filePath) {
   loadDefaultTSLint();
 
   if (config.useLocalTslint) {
-    return getLocalLinter(basedir);
+    return resolveAndCacheLinter(basedir);
+  }
+
+  if (config.useGlobalTslint) {
```
Comment (Xapphire13) [inline lib/worker.js]: Is `basedir` really used when using global tslint?

File: lib/worker.js (hunk 7 above)
Comment (smhxx) [reply lib/worker.js]: Yeah, I juggled with doing it this way vs. using `reject()` and opted for this mainly because the try/catch in `getLinter()` looked cleaner to me than the alternative "two-callbacks" method. I suppose it's really personal preference and I'll defer to you guys if you have a preference one way or the other. [It does work, though.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)

> If an error is thrown in the executor function, the promise is rejected. The return value of the executor is ignored.

Comment (Xapphire13) [reply lib/worker.js]: That's great! I totally agree it looks cleaner =]. I'll be adopting this style from now on I think.

File: lib/worker.js (hunk 8 above)
Comment (smhxx) [reply lib/worker.js]: It would also be an option to separately cache the global linter after the first time we fetch it, to avoid having to resolve it again for each new directory, but ultimately it would need to be inserted into the `tslintCache` Map for each directory that we've confirmed doesn't have a local install, just like on line 109 where the same is done for the packaged "fallback" install.

File: lib/worker.js (hunk 7 above)
Comment (Xapphire13) [reply lib/worker.js]: Wait, just looked at it a little closer, the `.exec()` callback isn't guaranteed to be in the executor function though right? It may have a different call stack.

Comment (smhxx) [reply lib/worker.js]: Y'know, you're right. It would *normally* work, but the fact that it's taking place in another process in this case means that needs to be a `Promise.catch()` rather than a regular `try/catch` block. I just tested it by changing `npm` to `nom` and with that change made, now it does catch the rejection properly. 👍 Will update in a second.

**Comment by smhxx** [comment]
Heh, so this has turned into more changes and refactoring than I intended, but:

1. I've confirmed that `npm get prefix` now fails gracefully, simply outputting a message to `console.warn()` and then using the packaged version of tslint, with no hard errors thrown.
2. I went ahead and made that optimization I mentioned earlier, whereby the "fallback" linter is only resolved once, and then stored in a variable for use any time future directories don't have a local `tslint` install. If the "use global" flag is set, and it is able to find the global `tslint` install, it assigns that as the fallback; if the "use global" flag isn't set, or if it wasn't able to find the global install the first time, then it uses the packaged version as the fallback instead.

File: lib/worker.js [hunk 9]
```diff
@@ -58,14 +60,8 @@ function shim(Linter) {
...
 
-function loadDefaultTSLint() {
-  if (!tslintDef) {
-    // eslint-disable-next-line import/no-dynamic-require
-    tslintDef = require(tslintModuleName).Linter;
-  }
-}
-
-async function getLocalLinter(basedir) {
+async function resolveAndCacheLinter(fileDir, moduleDir) {
```
Comment (Xapphire13) [inline lib/worker.js]: This doesn't need to be marked as `async` if no `await`s are done within it

Comment (Xapphire13) [reply lib/worker.js]: I think this is my last comment =]. @Arcanemagus, any objections after this?

Comment (smhxx) [reply lib/worker.js]: Ah, true, since all it's doing is immediately returning a Promise, it doesn't really need to be marked as async. I'll go ahead and clean up those unnecessary `async` declarations on the functions that do that.

File: package.json (hunk 1 above)
Comment (smhxx) [reply package.json]: Hah, I don't know why I didn't just do that to begin with. I guess I was a bit shy contributing to a new repo and trying to keep my "lines modified" count down. 😝 It's fixed!

File: lib/worker.js [hunk 10]
```diff
@@ -84,30 +80,78 @@ async function getLocalLinter(basedir) {
             linter = require('loophole').allowUnsafeNewFunction(() => require(linterPath).Linter);
           }
         } else {
-          linter = tslintDef;
+          return resolve();
```
Comment (Arcanemagus) [inline lib/worker.js]: :bug: You should just need to call `resolve()`, I'm not actually sure what `return resolve()` would do, I'm guessing that `resolve()` returns some value that makes that make sense as I've seen that elsewhere without issues, but the `return` isn't needed.

File: lib/worker.js [hunk 11]
```diff
@@ -84,30 +80,78 @@ async function getLocalLinter(basedir) {
...
         }
-        tslintCache.set(basedir, linter);
+        tslintCache.set(fileDir, linter);
```
Comment (Arcanemagus) [inline lib/worker.js]: ❓ Looks like @Xapphire13's earlier comment got marked resolved without being resolved:
Shouldn't this be `baseDir`?

File: lib/worker.js [hunk 12]
```diff
@@ -58,14 +60,8 @@ function shim(Linter) {
...
-function loadDefaultTSLint() {
-  if (!tslintDef) {
-    // eslint-disable-next-line import/no-dynamic-require
-    tslintDef = require(tslintModuleName).Linter;
-  }
-}
-
-async function getLocalLinter(basedir) {
+function resolveAndCacheLinter(fileDir, moduleDir) {
+  const basedir = moduleDir || fileDir;
```
Comment (Arcanemagus) [inline lib/worker.js]: 🐛 `baseDir` please 😉

File: lib/worker.js [hunk 13]
```diff
@@ -84,30 +80,78 @@ async function getLocalLinter(basedir) {
...
-        tslintCache.set(basedir, linter);
+        tslintCache.set(fileDir, linter);
         return resolve(linter);
       },
     );
   });
 }
 
+function getNodePrefixPath() {
+  return new Promise((fulfill, reject) => {
```
Comment (Arcanemagus) [inline lib/worker.js]: 🐎 This is _extrememely_ unlikely to change during runtime, cache it after the first (successful) call and return that.

Comment (Arcanemagus) [reply lib/worker.js]: ❓ Any particular reason you used `fulfill` instead of the standard `resolve`? I wouldn't block over it, but the non-standard nature of it requires a double take on the code.

File: lib/worker.js [hunk 14]
```diff
@@ -84,30 +80,78 @@ async function getLocalLinter(basedir) {
...
+    if (config.globalNodePath) {
+      const globalLint = await resolveAndCacheLinter(basedir, config.globalNodePath);
+      if (globalLint) {
+        fallbackLinter = globalLint;
+        return globalLint;
+      }
+    }
+
+    const globalLint = await getNodePrefixPath()
+      .then(prefix => resolveAndCacheLinter(basedir, prefix))
```
Comment (Arcanemagus) [inline lib/worker.js]: :bug: Why is this using `Promise` syntax? Nothing in here is a callback function that requires a `Promise` wrapper to work...

File: lib/worker.js [hunk 15]
```diff
@@ -84,30 +80,78 @@ async function getLocalLinter(basedir) {
...
+      .catch((err) => {
+        // eslint-disable-next-line no-console
+        console.warn(`Attempted to load global tslint, but \`npm get prefix\`
+failed. Falling back to the packaged version of tslint. You can specify
+your prefix manually in the settings or linter-tslint config file.\n
+The error message encountered was:\n\n${err.message}`);
+      });
+    if (globalLint) {
+      fallbackLinter = globalLint;
```
Comment (Arcanemagus) [inline lib/worker.js]: ❓ Should this be cached in `tslintCache`?

File: lib/worker.js (hunk 9 above)
Comment (Arcanemagus) [reply lib/worker.js]: Just as a note, I personally leave the `async` on there just to make it obvious to myself that the function needs to be `await`ed on where it gets called, not a huge issue though as you are perfectly
 right that it isn't needed 😉.

**Comment by Arcanemagus** [comment]
Well that's annoying, it posted all my pending review things when I made the single comment [here](https://github.com/AtomLinter/linter-tslint/pull/196#discussion_r140402873) 😞.

File: lib/worker.js (hunk 10 above)
Comment (Xapphire13) [reply lib/worker.js]: I noticed this too. There shouldn't be I'll effect, but for consistency I agree we should remove it

**Comment by Xapphire13** [comment]
@Arcanemagus, GitHub won't let me response to your comment about `async`. Isn't the presence of `async` without an `await` illegal in TypeScript? I know it is in C#. I felt like since this lint TypeScript, we should align with its rules =]. Though I'm also not passionate about it haha.

This begs the question, should this project be rewritten in TypeScript? =]

File: lib/worker.js (hunk 12 above)
Comment (smhxx) [reply lib/worker.js]: I wasn't sure how you wanted me to deal with that, since it previously *was* called `basedir`, apparently for the sake of brevity on line 71. I guess changing that line to `{ basedir: baseDir }` wouldn't be the end of the world.

File: lib/worker.js (hunk 10 above)
Comment (smhxx) [reply lib/worker.js]: Good call. I was doing it (and I suspect others do it) in order to kick execution back up the stack immediately, hence avoiding the caching operation on line 85... but that doesn't really need to be outside of the if-block anyway, so moving that inside and then getting rid of the `else` seems like a cleaner option.

File: lib/worker.js (hunk 11 above)
Comment (smhxx) [reply lib/worker.js]: Actually, that's correct as-is (at least, it's what I intended, "correct" is rather subjective 😛.) It specifically needs to be fileDir because the lookup on lines 110-113 searches the cache for the linter that was used the last time a source file *in that directory* was linted. If we use baseDir, which might actually be the user's `npm` directory, then it'll look up which linter to use every time and it will never actually get cached in a way that it can find later.

File: lib/worker.js (hunk 13 above)
Comment (smhxx) [reply lib/worker.js]: The npm prefix doesn't actually have to be cached because this function is never called more than once. If it ever gets to the point that it needs to grab the prefix (because there was no local linter, and the user set the "use global tslint" option but didn't explicitly give their prefix in the settings) then it immediately uses the prefix to grab the global linter. If it's able to resolve it, then it caches that one as the "fallback linter," and if it's not able to resolve it, it caches linter-tslint's version as the "fallback linter." Either way, this function should never be called again, because `getLinter()` will never get past line 124 once `fallbackLinter` is set. 👍

As for the fulfill/resolve thing, I'm not actually sure... I thought maybe I had seen it in linter-eslint or something and copied the style, but now I can't find it anywhere. Dunno where I got it from, to be honest. xD I'll just change it to `resolve`, I don't really care either way and resolve does seem more "standard," as you said.

File: lib/worker.js (hunk 14 above)
Comment (smhxx) [reply lib/worker.js]: I guess the `.then()` isn't 100% necessary, I just figured it was easier to chain the two promises and `await` them together, but since `resolveAndCacheLinter()` never rejects, I guess it doesn't really have to be before the `catch()`. It could also be done like:

```js
const prefix = await getNodePrefixPath().catch((err) => {
...
});
if (prefix) {
    const globalLint = await resolveAndCacheLinter(basedir, prefix);
    fallbackLinter = globalLint;
    return globalLint;
}
```

I guess it's just preference as to which makes the flow more obvious? I'd be okay with either.

File: lib/worker.js (hunk 15 above)
Comment (smhxx) [reply lib/worker.js]: I would probably leave `fallbackLinter` in its own variable, personally? I mean, you *could* put it into the map with everything else, but since it could technically be the global linter *or* the linter-tslint packaged version, that sorta makes finding it again a little complicated. This way, it's just "No local linter? Alright, use the other one." and you're done.

**Comment by Arcanemagus** [comment]
> Isn't the presence of async without an await illegal in TypeScript?

Fascinating, I wonder if that's an artifact of TypeScript's method of transpilation to ES5? It's not necessarily a bug in JavaScript, it's technically a bit of overhead though as initializing an `async function` is more expensive as a regular `function`.

> This begs the question, should this project be rewritten in TypeScript?

I'm not against it, I have no experience with it though 😛.

File: lib/worker.js (hunk 13 above)
Comment (Arcanemagus) [reply lib/worker.js]: Makes sense about the calling of `getNodePrefixPath`, thanks for clearing that up!

`resolve` is actually part of the standard documentation ([Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise)), but you _can_ use whatever name you want... almost everyone just uses the standard names though lol.

File: lib/worker.js (hunk 15 above)
Comment (Arcanemagus) [reply lib/worker.js]: Makes sense. 👍

File: lib/worker.js [hunk 16]
```diff
@@ -99,15 +110,46 @@ async function getLinter(filePath) {
...
+  if (config.useGlobalTslint) {
+    if (config.globalNodePath) {
+      const globalLint = await resolveAndCacheLinter(basedir, config.globalNodePath);
+      if (globalLint) {
+        fallbackLinter = globalLint;
+        return globalLint;
+      }
+    }
+
+    const prefix = await getNodePrefixPath().catch((err) => {
```
Comment (Arcanemagus) [inline lib/worker.js]: This should be in a `try`/`catch` block:

```js
let prefix
try {
  prefix = await getNodePrefixPath();
} catch (err) {
  // eslint-disable-next-line no-console
  console.warn(`Attempted to load global tslint, but \`npm get prefix\`
failed. Falling back to the packaged version of tslint. You can specify
your prefix manually in the settings or linter-tslint config file.\n
The error message encountered was:\n\n${err.message}`);
}
```

That's what I meant as part of the "`Promise` syntax", sorry for not making that clear 😉.

**Comment by Arcanemagus** [comment]
Sorry about being all over the place with the reviews, apparently I can't remember comments post pending review comments tonight.

File: lib/worker.js [hunk 17]
```diff
@@ -58,21 +60,15 @@ function shim(Linter) {
...
+function resolveAndCacheLinter(fileDir, moduleDir) {
+  const baseDir = moduleDir || fileDir;
   return new Promise((resolve) => {
     if (!requireResolve) {
       requireResolve = require('resolve');
     }
     requireResolve(
       tslintModuleName,
-      { basedir },
+      { basedir: baseDir },
```
[the remark this answers is not in the document]
Comment (Arcanemagus) [reply lib/worker.js]: @smhxx Oh, this was supposed to be in the requested changes below.

File: lib/worker.js (hunk 16 above)
Comment (smhxx) [reply lib/worker.js]: Ah! The `.catch()`. 😛 Now I understand. Should be easy since that the two promises are separated already.

**Comment by smhxx** [comment]
As for TypeScript, the compiler will complain if you have an `async` function with a return type that doesn't satisfy `Promise<T>`, but afaik it won't error if the callback of that Promise doesn't actually do any async. I guess they trust you enough that if your function is explicitly returning a Promise, you're probably smart enough to know whether you need it or not. 😛

Going from JavaScript -> TypeScript is pretty easy if you aren't doing any wonky wtf-level polymorphism that would make Douglas Crockford cry. Making the jump probably wouldn't be as bad as it sounds, although I suppose the question is "why," if it's just gonna have to be compiled back into JS for `apm` anyway. xD

**Comment by Arcanemagus** [comment]
> I suppose the question is "why," if it's just gonna have to be compiled back into JS for apm anyway. xD

Type checking is _definitely_ a good thing, with `flow` at least we can use `babel` which means the possibility of having Atom do the transpilation on install ([configuration](https://github.com/AtomLinter/linter-eslint/blob/1a2c594004591656c2c1541262cec465a6e3c52f/package.json#L189-L208)). I'm guessing that somebody has probably written a TypeScript transpiler plugin.

File: lib/worker.js (hunk 17 above)
Comment (Arcanemagus) [inline lib/worker.js]: Looks like this got lost in the mess of comments, since this name needs to be sent as an object into `requireResolve`, let's just keep it as `basedir` in the entire function. Sorry for going back and forth on this!

Comment (smhxx) [reply lib/worker.js]: woops! It's okay, I'll change it back. haha

**Comment by Xapphire13** [comment]
> Isn't the presence of async without an await illegal in TypeScript?

I was wrong, I think this may just be a C#-ism (which returns ``warning CS1998: Async block lacks `await' operator and will run synchronously``)

> I suppose the question is "why," if it's just gonna have to be compiled back into JS for apm anyway. xD

I would say for type checking and features that JS (and ES6/7) doesn't have yet. It may also yield a slight perf. increase since it would no longer be transpiled ES6+ running through babel at runtime. @arcanemagus, we could get the TypeScript to compile upon install using `scripts: {postinstall: "tsc"}` or similar in the `package.json` file right?

**Comment by smhxx** [comment]
I bet you could even do it as a `prepublishOnly` script if you wanted to, and have users just download the compiled JS straight from npm. Not sure what the trade-off is between that and doing it at install, but I guess it's an option...

**Comment by Xapphire13** [comment]
Looks like `apm` doesn't support prepublish =[ https://github.com/atom/apm/issues/372

**Comment by Arcanemagus** [comment]
> transpiled ES6+ running through babel at runtime.

Atom has a `compileCache`, when a file is first loaded it checks if it needs to be transpiled. If it does it does the work then and then stores the transpiled version (with source map) in the `compileCache`. Whenever the file is loaded after that it loads it instead from the `compileCache` directly.

tl;dr it only transpiles files once 😉.

> @Arcanemagus, we could get the TypeScript to compile upon install using scripts: {postinstall: "tsc"} or similar in the package.json file right?

Not quite (for the reason @smhxx found). If you want to use transpiled code you either have to store both the source and transpiled version in the repository... or use the `atomTranspilers` key with a compatible package to control how Atom builds files for the `compileCache`. [`atom-babel6-transpiler`](https://www.npmjs.com/package/atom-babel6-transpiler) is one written for `babel`. I don't know if anybody has bothered to create one for TypeScript yet though...

**Comment by smhxx** [comment]
@Arcanemagus Surprisingly, I looked and wasn't able to find one anywhere. Really surprises me since I'd imagine plenty of people are probably interested in being able to use TS to write Atom packages.

~70 LOC later, [the problem has been resolved](https://www.npmjs.com/package/atom-ts-transpiler). 😉

**Comment by Xapphire13** [comment]
@smhxx, awesome job! I was thinking about making something similar myself =].

**Comment by Arcanemagus** [comment]
Btw, Atom actually has native support for `.ts` files.... but uses Some positively ancient version of `typescript` for it so it's near useless. Nice job on creating `atom-ts-transpiler`!
